### DISCUSSION TITLE
Remove glib dependency?

### DISCUSSION THREAD

**Discussion by MrSapps** [FluidSynth/fluidsynth#899] [Ideas]
I am trying to use fluidsynth in a game engine. There exists "lite" versions that remove glib which makes it very simple to build, however these version are extremely old (1.1.6 era).

Would it ever be considered to have an option to build without depending on glib? Even if its a subset of the whole lib functionality?

**Comment by derselbst** [comment]
In your post I'm missing arguments explaining why you want to remove glib.

> Would it ever be considered to have an option to build without depending on glib?

See #847.

> Even if its a subset of the whole lib functionality?

Unlikely. I will not cripple the library.

**Reply by MrSapps** [reply]
Because glib is too big and bloated of a dependency in a small game engine, its bigger than the entire game engine, also it doesn't work on many embedded platforms and its difficult to compile on windows platform.

I am not sure what you mean by crippling the library? An option to allow a more modularized build so certain things can be skipped if the consumer of the library isn't using them doesn't sound like anything would be crippled, rather it would be more flexible.

What is the issue with using OS native threads instead of glib threads for instance? Or refactoring the code to allow easy replacement of glib in a fork?

**Reply by sykhro** [reply]
> What is the issue with using OS native threads instead of glib threads for instance?

Having to maintain code to do the things GLib already does for different platforms. Considering fluidsynth is used by people on WinXP and even earlier, re-implementing threading routines using that kind of old API is a big maintenance burden. 
GLib also provides atomic operations, which are currently unavailable on Windows and would need to be reimplemented from scratch, and some other things that are used in the LADSPA bits.

> Or refactoring the code to allow easy replacement of glib in a fork?

This is already doable. It's a matter of changing a few defines in fluidsynth_priv.h and some bits of the internal API.
You'll find that removing GLib only simplifies things to a certain extent: you have to replace it with something else, and there aren't many C libraries that do the same and are sane to build on Windows.

**Reply by MrSapps** [reply]
#847 it was mentioned that fluidsynth_priv.h can be further improved/refactored to make this easier, and there are some platform specific bits that are not using glib that could be and wrapped/abstracted into this header - I think this sounds like a nice idea

**Reply by derselbst** [reply]
> Because glib is too big and bloated of a dependency in a small game engine, its bigger than the entire game engine

Link it statically to get only what's needed.

> and its difficult to compile on windows platform

Using vcpkg it becomes a piece of cake.

> I am not sure what you mean by crippling the library? An option to allow a more modularized build so certain things can be skipped if the consumer of the library isn't using them doesn't sound like anything would be crippled, rather it would be more flexible.

fluidsynth already provides various options that let you "modularize" the build as you please. But glib is located in fluidsynth's core functionality. I.e. you would need to mess around in the lib itself, that's why I'm talking about crippling.

> also it doesn't work on many embedded platforms

Glib is the main reason that fluidsynth can run on Windows9x, Solaris, OS/2. Please provide evidence for those "many embedded platforms" where it doesn't work.

> What is the issue with using OS native threads instead of glib threads for instance?

Maintainability? I can neither test it nor am I willing to re-invent the wheel again and make a step back in the 90s where nobody used OSALs. You cannot seriously suggest to use native OS threads on all of those OSs above, can you? I have a life...

**Reply by sykhro** [reply]
> Link it statically to get only what's needed.

Unfortunately this can't be done on Windows (or at least, not anymore)

**Reply by MrSapps** [reply]
> Link it statically to get only what's needed.

This isn't possible on windows.
 
> Using vcpkg it becomes a piece of cake.

It doesn't. My program uses a static CRT and the vcpkg libs only use a dynamic CRT.

> fluidsynth already provides various options that let you "modularize" the build as you please. But glib is located in fluidsynth's core functionality. I.e. you would need to mess around in the lib itself, that's why I'm talking about crippling.

I don't understand why allowing the glib layer to be swapped out is "crippling" the library then?

> > also it doesn't work on many embedded platforms
> 
> Glib is the main reason that fluidsynth can run on Windows9x, Solaris, OS/2. Please provide evidence for those "many embedded platforms" where it doesn't work.
> 

Xbox 1, Playstation 1/2/3 and ESP MCU probably others.

> > What is the issue with using OS native threads instead of glib threads for instance?
> 
> Maintainability? I can neither test it nor am I willing to re-invent the wheel again and make a step back in the 90s where nobody used OSALs. You cannot seriously suggest to use native OS threads on all of those OSs above, can you? I have a life...

Sure I can see why that would be a problem. But allowing glib to be easily switched out can't hurt maintaince?

**Reply by sykhro** [reply]
> Xbox 1, Playstation 1/2/3 and ESP MCU probably others.

Realistically those can never be supported due to proprietary SDKs (or obscure homebrew toolchains). Besides, you'd still need to write an audio driver for each of them, which is no small feat compared to the little editing needed as of now to (hackily) swap out GLib

**Reply by derselbst** [reply]
> I don't understand why allowing the glib layer to be swapped out is "crippling" the library then?

You initially said: "Even if its a subset of the whole lib functionality?" in connection with "have an option to build without depending on glib" Since glib is used by fluidsynth's core functionality, I replied with "crippling".

> Xbox 1, Playstation 1/2/3 and ESP MCU probably others.

Not to forget Nintendo, Sega, Amiga and C64... Sorry, I cannot take this seriously.

> But allowing glib to be easily switched out can't hurt maintaince?

Have you ever looked at the code? The whole glib stuff is located in three files: `fluid_sys.c`, `fluid_sys.h` and `fluidsynth_priv.h` . So what exactly are you complaining about?

**Reply by MrSapps** [reply]
> > Xbox 1, Playstation 1/2/3 and ESP MCU probably others.
> 
> Realistically those can never be supported due to proprietary SDKs (or obscure homebrew toolchains). Besides, you'd still need to write an audio driver for each of them, which is no small feat compared to the little editing needed as of now to (hackily) swap out GLib

`fluid_synth_write_s16` is enough to make it work on Xbox1, PS2 and 3, PS1 needs a lot more edits so I probably shouldn't have mentioned that one.

> > I don't understand why allowing the glib layer to be swapped out is "crippling" the library then?
> 
> You initially said: "Even if its a subset of the whole lib functionality?" in connection with "have an option to build without depending on glib" Since glib is used by fluidsynth's core functionality, I replied with "crippling".

It usually implies removing functionality which wouldn't have been the case, so I was confused by that statement.

> > Xbox 1, Playstation 1/2/3 and ESP MCU probably others.
> 
> Not to forget Nintendo, Sega, Amiga and C64... Sorry, I cannot take this seriously.

Why can't you take this seriously? If Windows9X is supported the NT kernel in Xbox 1 years newer and the hardware is much more powerful than PC's of the Win9x era.

> > But allowing glib to be easily switched out can't hurt maintaince?
> 
> Have you ever looked at the code? The whole glib stuff is located in three files: `fluid_sys.c`, `fluid_sys.h` and `fluidsynth_priv.h` . So what exactly are you complaining about?

I will try getting it to use Windows OS calls and let you know how it goes. If there is some easy edits that makes this easier would they be accepted in a PR?

**Reply by derselbst** [reply]
> > > Xbox 1, Playstation 1/2/3 and ESP MCU probably others.
> > 
> > 
> > Not to forget Nintendo, Sega, Amiga and C64... Sorry, I cannot take this seriously.
> 
> Why can't you take this seriously?

Because it appears to me that you were just listing proprietary platforms where glib (and no other free, open, portable OSAL) is supported.

> If there is some easy edits that makes this easier would they be accepted in a PR?

Most likely yes.

**Reply by MrSapps** [reply]
I've added a PR here that pushes all of the glib calls into the priv header etc <https://github.com/FluidSynth/fluidsynth/pull/901> with this I can add glib_replacement.h in my fork and have it build without using glib
